cloudinit/sources/helpers/openstack.py

```diff
@@ -646,13 +660,28 @@ def convert_net_json(network_json=None, known_macs=None):
                     }
                 )
 
-            dns_nameservers = [
-                service["address"]
-                for service in network.get("services", [])
-                if service.get("type") == "dns"
-            ]
+            # Look for either subnet or network specific DNS servers
```

Comment:
As we are not going to deal with tons of dns nameservers, we can probably keep it simple using lists, maintaining the order in which the dns nameserves are defined in network_json by doing something in the lines of the following patch. WDYT?

```diff
diff --git a/cloudinit/sources/helpers/openstack.py b/cloudinit/sources/helpers/openstack.py
index 558f024e4..9b46a22c3 100644
--- a/cloudinit/sources/helpers/openstack.py
+++ b/cloudinit/sources/helpers/openstack.py
@@ -661,27 +661,23 @@ def convert_net_json(network_json=None, known_macs=None):
                 )
 
             # Look for either subnet or network specific DNS servers
-            # and add them as subnet level DNS entries. Use a set to
-            # for accumulation to eliminate duplicates.
+            # and add them as subnet level DNS entries.
             # Subnet specific nameservers
-            dns_nameservers = set(
-                [
-                    service["address"]
-                    for route in network.get("routes", [])
-                    for service in route.get("services", [])
-                    if service.get("type") == "dns"
-                ]
-            )
+            dns_nameservers = [
+                service["address"]
+                for route in network.get("routes", [])
+                for service in route.get("services", [])
+                if service.get("type") == "dns"
+            ]
             # Network specific nameservers
-            dns_nameservers.update(
-                [
-                    service["address"]
-                    for service in network.get("services", [])
-                    if service.get("type") == "dns"
-                ]
-            )
+            for service in network.get("services", []):
+                if service.get("type") != "dns":
+                    continue
+                if service["address"] in dns_nameservers:
+                    continue
+                dns_nameservers.append(service["address"])
             if dns_nameservers:
-                subnet["dns_nameservers"] = list(dns_nameservers)
+                subnet["dns_nameservers"] = dns_nameservers
 
             # Enable accept_ra for stateful and legacy ipv6_dhcp types
             if network["type"] in ["ipv6_dhcpv6-stateful", "ipv6_dhcp"]:
diff --git a/tests/unittests/sources/helpers/test_openstack.py b/tests/unittests/sources/helpers/test_openstack.py
index ec85cbeec..6ec0bd75b 100644
--- a/tests/unittests/sources/helpers/test_openstack.py
+++ b/tests/unittests/sources/helpers/test_openstack.py
@@ -340,7 +340,7 @@ class TestConvertNetJson:
                                 }
                             ],
                             "address": "x.x.x.x",
-                            "dns_nameservers": ["8.8.8.8", "1.1.1.1"],
+                            "dns_nameservers": ["1.1.1.1", "8.8.8.8"],
                             "ipv4": True,
                         }
                     ],

```

Reply:
Thanks for adding the test case, I had intended to do that but overlooked it when I was updating the other test cases.

I agree, simpler is better. I've pushed a new PR with the modifications you suggested.